### ISSUE TITLE
Define validity for a Service Base URL and define handling of invalid SBUs

### ISSUE THREAD

**Issue by jdickey** [jdickey/rack-service_api_versioning#14]
(Yes, we're defining an aspect of a feature, and it's a bug because it wasn't defined earlier.)

Consider an AVIDA whose Service Base URL (SBU) is `http://api.example.com:4321/` and which has an API Version implementation with its SBU specified as `http://api.example.com/path/to/apidemo`. Note the lack of a trailing `/` in that latter SBU definition.

Recall that when a request is made to an AVIDA, such as `http://api.example.com:4321/search?type=wombat`, the request, along with its path information, query parameters, and body content, if any, is to be redirected (via HTTP [307](https://httpstatuses.com/307)) to the SBU for the relevant API Version *along with the as-supplied path information, query parameters, and body content.* That *strongly* implies that the SBU should be a resource-type path, or a "directory", in form; i.e., that the last character of the SBU **must** be a `/`.  How do we handle cases where the SBU supplied when that API Version is Registered does not meet that condition? We essentially have three choices:

1. Append a trailing `/` to the end of the as-supplied SBU; the example would then become `http://api.example.com/path/to/apidemo/` and the full request URL as `http://api.example.com/path/to/apidemo/search?type=wombat`;
2. Truncate any text following a final `/` from the as-supplied SBU; the full request URL then becomes `http://api.example.com/path/to/search?type=wombat`;
3. Detect the invalid SBU, log it, and fail the request with an HTTP [400](https://httpstatuses.com/400) status code.

While option 3 would be the least ambiguous handling, it would also quite likely be the least [robust](https://en.wikipedia.org/wiki/Robustness_principle). Either of the first two options, if not matching the assumption made by the client developer, will lead to equally non-robust request failures, heralded by HTTP [404](https://httpstatuses.com/307)s.

While an initial Slack discussion argued for Option 2 (truncation), we are going to rethink that after a bit of sleep (check the timestamp of this issue opening). Why would someone supply a part of the URL that they never intended to use (which is really what Option 2 argues for)? Should we *just assume* that they accidentally omitted a trailing slash (per Option 1), or should we say "we can't read your mind; fix the data and try again" (Option 3)?

**Comment by jdickey** [comment]
And, after continuing to poke around, Issue #12 seems surprisingly resistant to quick workarounds not dependent on fixing this. More likely, we've developed tunnel vision by [focusing in too tightly](https://en.wikipedia.org/wiki/Blind_men_and_an_elephant) on what has already sprawled far more than is healthy.

Perhaps we should close PR #13 pending a fix here, then either reopen it or do a new "fix Issue #12" PR that has the effect of recreating Commits 5fa1389 and fa24b2f from PR #13.

**Comment by jdickey** [comment]
And, after letting this sift in the back of our mind while poking at other things for 24 hours, we've come back to Option 1 as the behaviour to be implemented. (And *documented!*) If you specify a path component, like `apidemo` in `http://api.example.com/path/to/apidemo`, then our assumption *has* to be that your intent was to include that as a segment *in the query path based on the SBU*; adding per-request actions to that _only_ makes sense if we proceed by appending a path-separator `/`, as described in Option 1 above.

Pffffffffft.

**Comment by jdickey** [comment]
A Service Base URL, or SBU, is by definition the _base_ for constructing a URL that routes to a specific action supported by a Service Component's API. We've occasionally used the example of an APIDemo Component Service with a `query` action; the SBU for its API Version-Independent Delivery Application, or AVIDA, might be `http://dunsel.example.com/apidemo/`. Content negotiation might resolve an API Version SBU of `http://services.dunsel.com:9876/v2/`.

**Note that** both of these SBUs include a final `/`. For a URL to be a valid SBU, it **must** end with a terminating slash. URLs *not* so terminated are *invalid,* and **must** cause an HTTP [400](https://httpstatuses.com/400) status code to terminate request processing. The HTTP response containing the 400 status code **should** include a response-body string reporting the invalid SBU and stating that it is invalid.

An example response body might be formatted as `Invalid formatting for Service Base URL: http://dunsel.example.com/apidemo`.

It could easily be argued that this is an insufficiently [robust](https://en.wikipedia.org/wiki/Robustness_principle) practice, and that simply adding a trailing `/` would be a better practice. Were this for a more public-facing API, we might well agree; however, for what was originally intended as a somewhat over-engineered internal API to begin with, enough effort flailing around with alternatives has already been invested. Onwards!

**Comment by jdickey** [comment]
The previous comment left open the question of _where_ an invalidly non-terminated SBU should fail the request. One *could* argue "as soon as practicable in the middleware flow", noting that since the SBU is first acquired as part of the data clump encoded by the `ServiceComponentDescriber` class, that that would be the logical place: at the first step of our three-step middleware chain.

Others would note that that middleware component doesn't "know" or care about the *details* of the data it acquires; only that an entity procured from a repository matches a Service Name. They might then note that the first mention in the normal flow of an SBU *as an SBU* is in the `Rack::ServiceApiVersioning::EncodedApiVersionData::ReturnData` class, where `:base_url` is an attribute of a `Dry::Struct::Value` class that can easily support validation such as checking that the string is properly terminated. We find this latter argument more compelling, as it takes greater advantage of existing code while continuing to guarantee that any AVIDA using the middleware correctly will still abort requests when presented with invalid data.

**Comment by jdickey** [comment]
We have a [temporary branch](https://github.com/jdickey/rack-service_api_versioning/tree/wtf-from-fix-12-api-redirect) with some new tests for verifying that an invalid SBU causes an HTTP 400 status code to be generated. Commenting out the tests for that, with the code in place, continues to yield a green bar. Enabling the tests causes failures — in apparently unrelated tests; the invalid-SBU-generates-`400` tests always pass.

As a reminder, the only known way to run MiniTest tests in a reproducible order requires a somewhat cumbersome syntax to set the randomisation seed; e.g

```
ruby -e 'require "./test/rack/service_api_versioning/accept_content_type_selector_test.rb"' -- -v --seed=16442
```

Using a seed value of `16442` will yield a green bar when used for Commit 8835a06 against either Ruby 2.3.3p222 or 2.4.0p0. Most other seeds, such as `42379`, will cause errors (*not* test failures).

The output from running tests with several different seeds has been moved to [a Gist](https://gist.github.com/jdickey/83f202e04f5f23e828132651e540f04c), where they are 
hopefully more legible and easily compared.

**Comment by jdickey** [comment]
Commit 82441e7 solves the problem; now to clean up the Issue #12 vs Issue #14 PR spaghetti.

**Comment by jdickey** [comment]
The former temporary branch has been renamed and now serves as the basis for PR#15. It will address both this and Issue #12.

**Comment by jdickey** [comment]
Also remember that actual function definitions (e.g., [`env_fixture`](https://github.com/jdickey/rack-service_api_versioning/blob/82441e7/test/rack/service_api_versioning/accept_content_type_selector_test.rb#L22-L24) in [`accept_content_type_selector_test.rb`](https://github.com/jdickey/rack-service_api_versioning/blob/82441e7/test/rack/service_api_versioning/accept_content_type_selector_test.rb)) are *shared among all MiniTest tests,* which will put us right back into a situation very much like the one we just left.

So how to memoise test-specific fixtures without introducing name collisions likely to break the code? Simple: use a (memoised, per-test-script) `let` fixture that is an object with class methods that generate fixtures.

Instead of, for example,

```ruby

def app_fixture
  # stuff
end

def env_fixture
  # stuff
end

# ...

it 'runs a test' do
  env = env_fixture
  # fiddle with `env` entries
  app = app_fixture env
  # use `app` and `env` to do stuff in tests
  expect(something).to_equal expected
end
```

we do something like

```ruby
let(:fixture_generator) do
  Class.new do
    def self.app_fixture
      # stuff
    end
    
    def self.env_fixture
      # stuff
    end
  end # no need to call `.new` here...
end

# ...

it 'runs a test' do
  env = fixture_generator.env_fixture
  # fiddle with `env` entries
  app = fixture_generator.app_fixture env
  # use `app` and `env` to do stuff in tests
  expect(something).to_equal expected
end
```
That gives us the fixture generation without shared state, without risking generator function name conflicts between multiple tests that each need generated `app` and `env` (and other) data. It also allows us to reuse fixture generators (between tests) when appropriate *without* sharing state *per se.* Win all the way 'round?

**Comment by jdickey** [comment]
One could also easily argue that this "rigamarole", this "defensive programming" *for test code* calls attention to a serious defect in either the test tool or the way it is being used: why does managing fixtures and state *in tests* require so much effort and thought? Test-driven development has been defined as "write the simplest test that can possibly fail, then write the simplest code that can make the test pass". We're going a long way away from that.

**Comment by jdickey** [comment]
Should have been marked closed by Commit f277920.

**Comment by jdickey** [comment]
Should have been marked closed by Commit f277920.
